File: packages/shared/src/handoff.ts
import { z } from "zod";

// --- Enums ---

export const HandoffStatus = z.enum(["pending", "completed", "failed"]);
export type HandoffStatus = z.infer<typeof HandoffStatus>;

// --- Schema ---

export const HandoffSchema = z.object({
  id: z.string().uuid(),
  fromAgentId: z.string().uuid(),
  toAgentId: z.string().uuid().nullable(),
  moduleId: z.string().uuid(),
  summary: z.string(),
  memorySnapshotId: z.string().uuid().nullable(),
  status: HandoffStatus,
});
export type Handoff = z.infer<typeof HandoffSchema>;

File: apps/hiveweave/lib/hiveweave/schema/handoff.ex
defmodule HiveWeave.Schema.Handoff do
  use Ecto.Schema
  import Ecto.Changeset

  schema "handoffs" do
    field :from_agent_id, :string
    field :to_agent_id, :string
    field :module_id, :string
    field :status, :string  # pending | accepted | completed | approved
    field :summary, :string
    field :expect_report, :integer, default: 0
    field :reported_up, :integer, default: 0
    field :created_at, :integer
    field :updated_at, :integer
    field :completed_at, :integer
  end

  def changeset(handoff, attrs) do
    handoff
    |> cast(attrs, [
      :from_agent_id, :to_agent_id, :module_id, :status, :summary,
      :expect_report, :reported_up, :created_at, :updated_at, :completed_at
    ])
    |> validate_required([:from_agent_id, :to_agent_id, :status, :created_at])
  end
end
